README.md
Google-DataStore version 0.01
=============================
   Google::DataStore - Perl Client Library for Google DataStore
   
   Provides access to "Google DataStore", this initial version allows:
   
   - Make lookup
   - Initialize transactions
   - Commit (missing many validations for including mutations, but it works)
   - Send a simple query with GqlQuery

# NAME
   Google::DataStore - Perl Client Library for Google DataStore

# INSTALLATION

To install this module type the following:

   * perl Makefile.PL
   * make
   * make test
   * make install

# DEPENDENCIES

This module requires these other modules and libraries:

# SYNOPSIS

    use Google::DataStore;
    my $client_email = <YOUR_CLIENT_EMAIL_ADDRESS>;
    my $private_key_file = <YOUR_PRIVATE_KEY_FILE>;
    my $project_id = <YOUR_PROJECT_ID>;

    my $datastore = Google::DataStore::create(
       client_email => $client_email,
       private_key_file => $private_key_file,
       project_id => $project_id,
    );
   
    #define the dataset to use
    $datastore->use_dataset('<YOUR_DATASET_ID>');

    #let's do a simple Lookup
    #create a new key
    my $key = $datastore->Key();
    #Add the path element to the key
    my $path = $key->AddPathElement(kind=>'Trivia', name=>'hgtg');
   
    #the lookup
    my $resp = $datastore->lookup(keys=>$key->{key});
    if (defined $resp->{found}[0]){
       my $entity = $resp->{found}[0]{entity};
       #This is only to see the entity
       use Dumper::Data;
       print Dumper($entity);
    }

# COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2015 by Omar Yesith Alvarado González - omarjesith@gmail.com

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.4 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Google::DataStore.

0.01  Thu Nov 19 15:45:05 2015
	- original version; created by h2xs 1.23 with options
		-AX -n Google::DataStore



lib/Google/DataStore/Key.pm
package Google::DataStore::Key;
use 5.010001;
use strict;
use warnings;


sub new {
  my ($class, %args) = @_;
  my $self = bless { %args }, $class;

  $self->{key}{partitionId}{datasetId} = $args{datasetId} if defined $args{datasetId};
  $self->{key}{partitionId}{namespace} = $args{namespace} if defined $args{namespace};

  return $self;
}

sub DESTROY {
}

sub AddPathElement {
  my ($self, %args) = @_;
  my @path = {};
  push @{$self->{key}{path}}, @path;

  $self->{key}{path}[scalar(@{$self->{key}{path}})-1]{kind} = $args{kind} if defined $args{kind};
  $self->{key}{path}[scalar(@{$self->{key}{path}})-1]{name} = $args{name} if defined $args{name};
  $self->{key}{path}[scalar(@{$self->{key}{path}})-1]{id} = $args{id} if defined $args{id};

  return $self->{key}{path}[scalar(@{$self->{key}{path}})-1];
}
1;


lib/Google/DataStore/V1beta2.pm
package Google::DataStore::V1beta2;
use 5.010001;
use strict;
use warnings;

use base qw(Google::DataStore);

use HTTP::Request;
use JSON qw(decode_json encode_json);
use URI::Escape;

sub new {
  my ($class, %args) = @_;

  my $self = $class->SUPER::new(
    %args,
    GOOGLE_DATASTORE_REST_DESCRIPTION => 'https://www.googleapis.com/discovery/v1/apis/datastore/v1beta2/rest',
    GOOGLE_DATASTORE_API_BASE_URL => 'https://www.googleapis.com/datastore/v1beta2/datasets'
  );

  bless $self, $class;
}

sub DESTROY {
}

sub request {
  my ($self, %args) = @_;

  my $resource = $args{resource} || die;
  my $method = $args{method} || die;
  my $project_id = defined $args{project_id} ? $args{project_id} : $self->{project_id};
  my $dataset_id = defined $args{dataset_id} ? $args{dataset_id} : $self->{dataset_id};
  my $table_id = $args{table_id} || '';
  my $job_id = $args{job_id} || '';

  my $header = HTTP::Headers->new(Authorization => "Bearer $self->{access_token}{access_token}");
  my $rest_description = $self->{rest_description}{resources}{$resource}{methods}{$method} || die;
  my $http_method = $rest_description->{httpMethod};
  my $path = join('/', $self->{GOOGLE_DATASTORE_API_BASE_URL}, $rest_description->{path});
  $path =~ s/{projectId}/$project_id/;
  $path =~ s/{tableId}/$table_id/;
  $path =~ s/{datasetId}/$dataset_id/;
  $path =~ s/{jobId}/$job_id/;
  if ($self->{exp} < time + 60) {
    $self->_auth;
  }
  
  my $request;
  if ($resource eq 'jobs' && $method eq 'insert') {
    my $sub_method = (keys %{$args{content}->{configuration}})[0];

    if ($sub_method !~ /^load$/ || defined $args{content}->{configuration}{load}{sourceUris}) {
      $request = HTTP::Request->new($http_method, $path, $header);
      if ($http_method =~ /^(?:POST|PUT|PATCH)$/) {
        $request->header('Content-Type' => 'application/json');
        $request->content(encode_json($args{content}));
      }
    } else {
      my $upload_path = $rest_description->{mediaUpload}{protocols}{simple}{path};
      $upload_path = "https://www.googleapis.com" . $upload_path;
      $upload_path =~ s/{projectId}/$project_id/;
      $upload_path .= "?uploadType=multipart";

      $header->header('Content-Type' => 'multipart/related');
      $request = HTTP::Request->new($http_method, $upload_path, $header);

      $request->add_part(
        HTTP::Message->new(
          ['Content-Type' => 'application/json; charset=UTF-8'],
          encode_json($args{content})
        )
      );

      my $data;
      open my $in, "<", $args{data} or die "can't open $args{data} : $!";
      $data = join('', <$in>);
      close $in;

      $request->add_part(
        HTTP::Message->new(
          ['Content-Type' => 'application/octet-stream'],
          $data
        )
      );
    }

    my $response = $self->{ua}->request($request);
    if (defined $response->content) {
      my $content = decode_json($response->content);
      if (defined $content->{error}) {
        return { error => { message => $content->{error}{message} } };
      } elsif ($response->code == 200) {
        my $json_response = decode_json($response->decoded_content);

        # return if async is true
        return $json_response if $args{async};

        my $job_id = $json_response->{jobReference}{jobId};
        while (1) {
          my $json_response = $self->request(method => 'get', resource => 'jobs', job_id => $job_id);
          if ($json_response->{status}{state} eq 'DONE') {
            return $json_response;
          } else {
            print "Wating...(state: $json_response->{status}{state})\n" if defined $self->{verbose};
            sleep(1);
          }
        }
      } else {
        die;
      }
    } else {
      die;
    }
  } else {
    if (defined $args{query_string} && %{$args{query_string}}) {
      my @query_string;
      while (my ($key, $value) = each %{$args{query_string}}) {
        push @query_string, join('=', uri_escape($key), uri_escape($value));
      }
      $path = join('?', $path, join('&', @query_string));
    }

    my $request = HTTP::Request->new($http_method, $path, $header);
    if ($http_method =~ /^(?:POST|PUT|PATCH)$/) {
      $request->header('Content-Type' => 'application/json');
      $request->content(encode_json($args{content}));
    }
    my $response = $self->{ua}->request($request);

    if ($response->code == 204) {
      return {};
    } elsif (defined $response->content) {
      # easy json check
      if ($response->content =~ /^\s*[{\[]/) {
        return decode_json($response->decoded_content);
      } else {
        return { error => { message => $response->content }};
      }
    } else {
      return { error => { message => 'Unknown Error' }};
    }
  }
}

1;


lib/Google/DataStore/CommitRequest.pm
package Google::DataStore::CommitRequest;
use 5.010001;
use strict;
use warnings;


sub new {
  my ($class, %args) = @_;
  my $self = bless { %args }, $class;
  return $self;
}

sub transaction {
  my ($self, $transaction) = @_;
  $self->{request}{transaction} = $transaction // return;
}

sub AddMutation {
	my ($self, %args) = @_;

	die "undefined mutation to set" if !defined $args{mutation};

	$self->{mutation} = $args{mutation};
	$self->{request}{mutation}{$self->{mutation}} = [];
	$self->{request}{mutation}{$self->{mutation}}[0]{key} = $args{key} if defined $args{key};	
	$self->{request}{mutation}{$self->{mutation}}[0]{properties} = {};
}

sub AddKey {
	my ($self, $key) = @_;

	die "undefined key to attach" if !defined $key;

	$self->{request}{mutation}{$self->{mutation}}[0]{key} = $key;
}

sub AddProperty {
	my ($self, %args) = @_;

	die "undefined property name" if !defined $args{name};
	die "undefined property type" if !defined $args{type};
	die "undefined property value" if !defined $args{value};
	
	$self->{request}{mutation}{$self->{mutation}}[0]{properties}{$args{name}}{$args{type}} = $args{value};
}
1;

lib/Google/DataStore.pm
package Google::DataStore;
use 5.010001;
use strict;
use warnings;

our $VERSION = '0.01';

use Class::Load qw(load_class);
use Crypt::OpenSSL::PKCS12;
use JSON qw(decode_json encode_json);
use JSON::WebToken;
use LWP::UserAgent;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Google::DataStore ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

sub create {
  my (%args) = @_;

  my $version = $args{version} // 'v1beta2';
  my $class = 'Google::DataStore::' . ucfirst($version);

  if (load_class($class)) {
    return $class->new(%args);
  } else {
    die "Can't load class: $class";
  }
}

sub new {
  my ($class, %args) = @_;

  die "undefined client_email" if !defined $args{client_email};
  die "undefined private_key_file" if !defined $args{private_key_file};
  die "not found private_key_file" if !-f $args{private_key_file};

  my $self = bless { %args }, $class;

  $self->{GOOGLE_API_TOKEN_URI} = 'https://accounts.google.com/o/oauth2/token';
  $self->{GOOGLE_API_GRANT_TYPE} = 'urn:ietf:params:oauth:grant-type:jwt-bearer';

  if ($self->{private_key_file} =~ /\.json$/) {
    open my $in, "<", $self->{private_key_file} or die "can't open $self->{private_key_file} : $!";
    my $private_key_json = decode_json(join('', <$in>));
    close $in;
    $self->{private_key} = $private_key_json->{private_key};
  } elsif ($self->{private_key_file} =~ /\.p12$/) {
    my $password = "notasecret";
    my $pkcs12 = Crypt::OpenSSL::PKCS12->new_from_file($self->{private_key_file});
    $self->{private_key} = $pkcs12->private_key($password);
  } else {
    die "invalid private_key_file format";
  }

  $self->_auth;
  $self->_set_rest_description;

  return $self;
}

sub DESTROY {
}

sub _auth {
  my ($self) = @_;

  $self->{scope} //= [qw(https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/datastore)];
  $self->{exp} = time + 3600;
  $self->{iat} = time;
  $self->{ua} = LWP::UserAgent->new;

  my $claim = {
    iss => $self->{client_email},
    scope => join(" ", @{$self->{scope}}),
    aud => $self->{GOOGLE_API_TOKEN_URI},
    exp => $self->{exp},
    iat => $self->{iat},
  };

  my $jwt = JSON::WebToken::encode_jwt($claim, $self->{private_key}, 'RS256', { type => 'JWT' });

  my $response = $self->{ua}->post(
    $self->{GOOGLE_API_TOKEN_URI},
    { grant_type => $self->{GOOGLE_API_GRANT_TYPE}, assertion => $jwt }
  );
  
if ($response->is_success) {
    $self->{access_token} = decode_json($response->decoded_content);
  } else {
    my $error = decode_json($response->decoded_content);
    die $error->{error};
  }
}

sub _set_rest_description {
  my ($self) = @_;
  my $response = $self->{ua}->get($self->{GOOGLE_DATASTORE_REST_DESCRIPTION});
  $self->{rest_description} = decode_json($response->decoded_content);
}

sub use_project {
  my ($self, $project_id) = @_;
  $self->{project_id} = $project_id // return;
}

sub use_dataset {
  my ($self, $dataset_id) = @_;
  $self->{dataset_id} = $dataset_id // return;
}

sub use_namespace {
  my ($self, $namespace) = @_;
  $self->{namespace} = $namespace // return;
}

sub allocateIds {

}

sub beginTransaction {
  my ($self, %args) = @_;

  my $dataset_id = $args{dataset_id} // $self->{dataset_id};
  my $project_id = $args{project_id} // $self->{project_id};

  my $isolationLevel = (defined $args{isolationLevel}) ? $args{isolationLevel} : 'SNAPSHOT';
  my $content->{isolationLevel} = $isolationLevel;

  my $response = $self->request(
    resource => 'datasets',
    method => 'beginTransaction',
    project_id => $project_id,
    dataset_id => $dataset_id,
    content => $content,
  );
  $self->{response} = $response;

  if (defined $response->{error}) {
    warn $response->{error}{message};
    return $response->{error}{message};
  } else {
    return $response->{transaction};
  }
}

sub CommitRequest {
  my ($self, %args) = @_;
  my $class = 'Google::DataStore::CommitRequest';
  
  if (load_class($class)) {
    return $class->new();
  } else {
    die "Can't load class: $class";
  } 
}

sub commit {
  my ($self, $content, %args) = @_;

  my $dataset_id = $args{dataset_id} // $self->{dataset_id};

  my $response = $self->request(
    resource => 'datasets',
    method => 'commit',
    dataset_id => $dataset_id,
    content => $content,
  );

  $self->{response} = $response;

  if (defined $response->{error}) {
    warn $response->{error}{message};
    return $response->{error}{message};
  } else {
    return $response;
  }

}

sub lookup {
  my ($self, %args) = @_;

  die "undefined keys to lookup" if !defined $args{keys};
  die "cannot specify both a read consistency and a transaction" if ( (defined $args{readConsistency}) && (defined $args{transaction}) );

  my $dataset_id = $args{dataset_id} // $self->{dataset_id};
  my $project_id = $args{project_id} // $self->{project_id};
  my @keys = $args{keys};

  my $content = {
    keys => \@keys
  };

  # options
  $content->{readOptions}{readConsistency} = $args{readConsistency} if defined $args{readConsistency};
  $content->{readOptions}{transaction} = $args{transaction} if defined $args{transaction};

  my $response = $self->request(
    resource => 'datasets',
    method => 'lookup',
    project_id => $project_id,
    dataset_id => $dataset_id,
    content => $content
  );
  $self->{response} = $response;

  if (defined $response->{error}) {
    warn $response->{error}{message};
    return $response->{error}{message};
  } else {
    return $response;
  }  
}

sub Key {
  my ($self, %args) = @_;
  my $class = 'Google::DataStore::Key';
	
  if (load_class($class)) {
    return $class->new();
  } else {
    die "Can't load class: $class";
  }  
}

sub rollback {

}

sub gqlQuery {
  my ($self, $query, %args) = @_;

  my $dataset_id = $args{dataset_id} // $self->{dataset_id};

  die "undefined query to run" if !defined $query;

  my $content->{gqlQuery}{queryString} = $query;

  my $response = $self->request(
    resource => 'datasets',
    method => 'runQuery',
    dataset_id => $dataset_id,
    content => $content
  );
  $self->{response} = $response;

  if (defined $response->{error}) {
    warn $response->{error}{message};
    return $response->{error}{message};
  } else {
    return $response->{batch}{entityResults};
  }  
}
1;
__END__

=head1 NAME

Google::DataStore - Perl extension for connect and interact with Google DataStore

=head1 SYNOPSIS

  use Google::DataStore;
  my $client_email = <YOUR_CLIENT_EMAIL_ADDRESS>;
  my $private_key_file = <YOUR_PRIVATE_KEY_FILE>;
  my $project_id = <YOUR_PROJECT_ID>;

  my $datastore = Google::DataStore::create(
     client_email => $client_email,
     private_key_file => $private_key_file,
     project_id => $project_id,
  );
  
  #define the dataset to use
  $datastore->use_dataset('<YOUR_DATASET_ID>');

  #let's do a simple Lookup
  #create a new key
  my $key = $datastore->Key();
  #Add the path element to the key
  my $path = $key->AddPathElement(kind=>'Trivia', name=>'hgtg');
   
  #the lookup
  my $resp = $datastore->lookup(keys=>$key->{key});
  if (defined $resp->{found}[0]){
     my $entity = $resp->{found}[0]{entity};
     #This is only to see the entity
     use Dumper::Data;
     print Dumper($entity);
  }

=head1 DESCRIPTION

Allows you to communicate and interact with Google Cloud Store, development version, allows authentication and execute simple processes such as:

- lookup
- transactions
- commit
- Queries with GqlQuery

TODO:
- Improve the management of the transactionRequest to validate the management of mutations correctly.
- Implement the rollback
- Extend the functionality of runQuery, currently only supports queries GqlQuery

=head2 EXPORT

None by default.


=head1 SEE ALSO

Updated source code at: https://github.com/omaryesith/Google-DataStore

=head1 AUTHOR

Omar Yesith Alvarado González, E<lt>omarjesith@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Omar Yesith Alvarado González

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.4 or,
at your option, any later version of Perl 5 you may have available.


=cut
